File: app/src/main/java/com/pr0gramm/app/feed/FeedItem.java
package com.pr0gramm.app.feed;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.services.HasThumbnail;

import org.joda.time.Instant;

/**
 * This is an item in pr0gramm feed item to be displayed. It is backed
 * by the data of an {@link Api.Feed.Item}.
 */
public class FeedItem implements Parcelable, HasThumbnail {
    private final Instant created;
    private final String thumb;
    private final String image;
    private final String fullsize;
    private final String user;
    private final int id;
    private final int promotedId;
    private final int width, height;
    private final short up;
    private final short down;
    private final byte mark;
    private final byte flags;
    private final boolean audio;

    public FeedItem(Api.Feed.Item item) {
        id = (int) item.getId();
        promotedId = (int) item.getPromoted();
        thumb = item.getThumb();
        image = item.getImage();
        fullsize = item.getFullsize();
        user = item.getUser();
        up = (short) item.getUp();
        down = (short) item.getDown();
        mark = (byte) item.getMark();
        created = item.getCreated();
        flags = (byte) item.getFlags();
        width = item.width().or(0);
        height = item.height().or(0);
        audio = item.audio().or(false);
    }

    @Override
    public long id() {
        return id;
    }

    public long promotedId() {
        return promotedId;
    }

    @Override
    @NonNull
    public String thumbnail() {
        return thumb;
    }

    public String image() {
        return image;
    }

    public String fullsize() {
        return fullsize;
    }

    public String user() {
        return user;
    }

    public int up() {
        return up;
    }

    public int down() {
        return down;
    }

    public int mark() {
        return mark;
    }

    public Instant created() {
        return created;
    }

    public int flags() {
        return flags;
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    public boolean audio() {
        return audio;
    }

    /**
     * Returns the content type of this Item, falling back to {@link ContentType#SFW}
     * if no type is available.
     */
    public ContentType contentType() {
        return ContentType.valueOf(flags).or(ContentType.SFW);
    }

    /**
     * Gets the id of this feed item depending on the type of the feed..
     *
     * @param type The type of feed.
     */
    public long id(FeedType type) {
        return type == FeedType.PROMOTED ? promotedId : id;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        // combine up/down as rating.
        int rating = ((up << 16) & 0xffff0000) | (down & 0xffff);

        dest.writeInt(this.id);
        dest.writeInt(this.promotedId);
        dest.writeString(this.thumb);
        dest.writeString(this.image);
        dest.writeString(this.fullsize);
        dest.writeString(this.user);
        dest.writeInt(rating);
        dest.writeByte(mark);
        dest.writeInt((int) (created.getMillis() / 1000));
        dest.writeByte(this.flags);
        dest.writeInt(width);
        dest.writeInt(height);
        dest.writeByte(audio ? (byte) 1 : 0);
    }

    FeedItem(Parcel in) {
        this.id = in.readInt();
        this.promotedId = in.readInt();
        this.thumb = in.readString();
        this.image = in.readString();
        this.fullsize = in.readString();
        this.user = in.readString();
        int rating = in.readInt();
        this.mark = in.readByte();
        this.created = new Instant(1000L * in.readInt());
        this.flags = in.readByte();
        this.width = in.readInt();
        this.height = in.readInt();
        this.audio = in.readByte() != 0;

        // extract up/down from rating
        this.up = (short) ((rating >> 16) & 0xffff);
        this.down = (short) (rating & 0xffff);
    }

    public boolean isVideo() {
        return image.endsWith(".webm") || image.endsWith(".mp4");
    }

    public static final Parcelable.Creator<FeedItem> CREATOR = new Parcelable.Creator<FeedItem>() {
        public FeedItem createFromParcel(Parcel source) {
            return new FeedItem(source);
        }

        public FeedItem[] newArray(int size) {
            return new FeedItem[size];
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/feed/FeedItem.kt
package com.pr0gramm.app.feed

import android.os.Parcel
import com.pr0gramm.app.Instant
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.listOfSize
import com.pr0gramm.app.parcel.ConstructorCreator
import com.pr0gramm.app.parcel.DefaultParcelable
import com.pr0gramm.app.parcel.javaClassOf
import com.pr0gramm.app.parcel.read
import com.pr0gramm.app.parcel.readStringNotNull
import com.pr0gramm.app.parcel.write
import com.pr0gramm.app.util.VideoQuality

/**
 * This is an item in pr0gramm feed item to be displayed. It is backed
 * by the data of an [Api.Feed.Item].
 */
data class FeedItem(
    val created: Instant,
    val thumbnail: String,
    val path: String,
    val fullsize: String,
    val user: String,
    val userId: Long,
    val id: Long,
    val promotedId: Long,
    val width: Int,
    val height: Int,
    val up: Int,
    val down: Int,
    val mark: Int,
    val flags: Int,
    val audio: Boolean,
    val deleted: Boolean,
    val variants: List<Api.Feed.Variant>,
    val subtitles: List<Api.Feed.Subtitle>,
    val placeholder: Boolean,
) : DefaultParcelable {
    constructor(item: Api.Feed.Item) : this(
        id = item.id,
        promotedId = item.promoted,
        userId = item.userId,
        thumbnail = item.thumb,
        path = item.image,
        fullsize = item.fullsize,
        user = item.user,
        up = item.up,
        down = item.down,
        mark = item.mark,
        created = item.created,
        flags = item.flags,
        width = item.width,
        height = item.height,
        audio = item.audio,
        deleted = item.deleted,
        variants = item.variants,
        subtitles = item.subtitles,
        placeholder = false,
    )

    /**
     * Returns the content type of this Item, falling back to [ContentType.SFW]
     * if no type is available.
     */
    val contentType: ContentType
        get() = ContentType.valueOf(flags) ?: ContentType.SFW

    val isVideo: Boolean
        get() = isVideoUri(path)

    val isImage: Boolean
        get() = isImageUri(path)

    val isPinned: Boolean
        get() = promotedId > 1_000_000_000

    /**
     * Gets the id of this feed item depending on the type of the feed..

     * @param type The type of feed.
     */
    fun id(type: FeedType): Long {
        return (if (type === FeedType.PROMOTED) promotedId else id)
    }

    fun pickVariant(quality: VideoQuality, mobile: Boolean, compatible: Boolean): Api.Feed.Variant {
        if (compatible) {
            // fallback to the "h264" or the default path
            return variants.firstOrNull { v -> v.name == "h264" }
                ?: Api.Feed.Variant(name = "base", path = path)
        }

        if (mobile && quality == VideoQuality.Adaptive) {
            val variant = variants.firstOrNull { v -> v.name == "vp9s" }
            if (variant != null) {
                return variant
            }
        }

        if (quality == VideoQuality.Adaptive || quality == VideoQuality.High) {
            val variant = variants.firstOrNull { v -> v.name == "vp9m" }
            if (variant != null) {
                return variant
            }
        }

        return variants.firstOrNull { v -> v.name == "vp9" }
            ?: variants.firstOrNull { v -> v.name == "h264" }
            ?: Api.Feed.Variant(name = "base", path = path)
    }

    override fun toString(): String = "FeedItem(id=$id)"

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id.toInt())
        dest.writeInt(promotedId.toInt())
        dest.writeInt(userId.toInt())

        // deduplicate values for thumb & fullsize if equal to 'image'
        dest.writeString(path)
        dest.writeString(thumbnail.takeIf { it != path })
        dest.writeString(fullsize.takeIf { it != path })

        dest.writeString(user)

        dest.writeInt(up)
        dest.writeInt(down)
        dest.write(created)
        dest.writeInt(width)
        dest.writeInt(height)

        var bits = 0

        bits = bits.or(if (audio) 0b0001 else 0)
        bits = bits.or(if (deleted) 0b0010 else 0)
        bits = bits.or(if (placeholder) 0b0100 else 0)

        bits = bits.or((mark and 0xff) shl 8)
        bits = bits.or((flags and 0xff) shl 16)

        dest.writeInt(bits)

        dest.writeByte(variants.size.coerceAtMost(16).toByte())
        for (variant in variants.take(16)) {
            dest.writeString(variant.name)
            dest.writeString(variant.path)
        }

        dest.writeByte(subtitles.size.coerceAtMost(16).toByte())
        for (variant in subtitles.take(16)) {
            dest.writeString(variant.language)
            dest.writeString(variant.path)
        }
    }

    companion object CREATOR : ConstructorCreator<FeedItem>(javaClassOf(), { source ->
        val id = source.readInt().toLong()
        val promotedId = source.readInt().toLong()
        val userId = source.readInt().toLong()

        // deduplicate values for thumb & fullsize if equal to image
        val image = source.readStringNotNull()
        val thumbnail = source.readString() ?: image
        val fullsize = source.readString() ?: image

        val user = source.readStringNotNull()

        val up = source.readInt()
        val down = source.readInt()
        val created = source.read(Instant)
        val width = source.readInt()
        val height = source.readInt()

        val bits = source.readInt()

        val audio = (bits and 0b0001) != 0
        val deleted = (bits and 0b0010) != 0
        val placeholder = (bits and 0b0100) != 0

        val mark = (bits ushr 8) and 0xff
        val flags = (bits ushr 16) and 0xff

        val variantCount = source.readByte().toInt()

        val variants = listOfSize(variantCount) {
            Api.Feed.Variant(
                name = source.readStringNotNull(),
                path = source.readStringNotNull(),
            )
        }

        val subtitleCount = source.readByte().toInt()
        val subtitles = listOfSize(subtitleCount) {
            Api.Feed.Subtitle(
                language = source.readStringNotNull(),
                path = source.readStringNotNull(),
            )
        }

        FeedItem(
            id = id,
            promotedId = promotedId,
            userId = userId,
            path = image,
            thumbnail = thumbnail,
            fullsize = fullsize,
            user = user,
            up = up,
            down = down,
            created = created,
            width = width,
            height = height,
            mark = mark,
            flags = flags,
            audio = audio,
            deleted = deleted,
            placeholder = placeholder,
            subtitles = subtitles,
            variants = variants,
        )
    })
}

fun isVideoUri(image: String): Boolean {
    return image.endsWith(".webm") || image.endsWith(".mp4")
}

fun isImageUri(image: String): Boolean {
    return image.endsWith(".jpg") || image.endsWith(".png")
}
